src/binary/decode.rs

```diff
@@ -0,0 +1,2179 @@
+    /// The output of decoding `Self` from the input stream.
+    ///
+    /// Since the custom section's arenas and AST nodes typically live inside
+    /// the `cx`, this is usually some sort of typed ID pointing into an arena.
+    type Output;
+
+    /// Decode an instance of `Self` into the given `cx`.
+    fn decode<Cx, R>(cx: &mut Cx, r: &mut R) -> Result<Self::Output, failure::Error>
+    where
+        Cx: Ids,
+        R: Read;
+}
+
+// Factored out to make testing easier.
+pub(crate) trait Ids {
+    fn webidl_bindings(&mut self) -> &mut WebidlBindings;
+    fn webidl_type_id(&self, index: u32) -> Result<Id<WebidlCompoundType>, failure::Error>;
+    fn binding_id(&self, index: u32) -> Result<Id<FunctionBinding>, failure::Error>;
+    fn wasm_func_id(&self, index: u32) -> Result<walrus::FunctionId, failure::Error>;
+    fn wasm_func_type_id(&self, index: u32) -> Result<walrus::TypeId, failure::Error>;
+}
+
+pub(crate) struct DecodeContext<'a> {
+    ids: &'a walrus::IndicesToIds,
+    pub(crate) webidl_bindings: WebidlBindings,
+}
+
+impl<'a> DecodeContext<'a> {
+    pub(crate) fn new(ids: &'a walrus::IndicesToIds) -> Self {
+        let webidl_bindings = WebidlBindings::default();
+        DecodeContext {
+            ids,
+            webidl_bindings,
+        }
+    }
+}
+
+impl Ids for DecodeContext<'_> {
+    fn webidl_bindings(&mut self) -> &mut WebidlBindings {
+        &mut self.webidl_bindings
+    }
+
+    fn webidl_type_id(&self, index: u32) -> Result<Id<WebidlCompoundType>, failure::Error> {
+        self.webidl_bindings
+            .types
+            .by_index(index)
+            .ok_or_else(|| failure::format_err!("no Web IDL type for index {}", index))
+    }
+
+    fn binding_id(&self, index: u32) -> Result<Id<FunctionBinding>, failure::Error> {
+        self.webidl_bindings
+            .bindings
+            .by_index(index)
+            .ok_or_else(|| failure::format_err!("no function binding for index {}", index))
+    }
+
+    fn wasm_func_id(&self, index: u32) -> Result<walrus::FunctionId, failure::Error> {
+        self.ids.get_func(index)
+    }
+
+    fn wasm_func_type_id(&self, index: u32) -> Result<walrus::TypeId, failure::Error> {
+        self.ids.get_type(index)
+    }
+}
+
+trait ReadExt: Read {
+    fn read_byte(&mut self) -> Result<u8, failure::Error>;
+    fn expect_byte(&mut self, expected: u8) -> Result<(), failure::Error>;
+    fn uleb(&mut self) -> Result<u32, failure::Error>;
+    fn ileb(&mut self) -> Result<i32, failure::Error>;
+    fn vec<T, Cx, E>(&mut self, cx: &mut Cx, e: &mut E) -> Result<(), failure::Error>
+    where
+        T: Decode,
+        Cx: Ids,
+        E: Extend<<T as Decode>::Output>;
+    fn option<T, Cx>(
+        &mut self,
+        cx: &mut Cx,
+    ) -> Result<Option<<T as Decode>::Output>, failure::Error>
+    where
+        T: Decode,
+        Cx: Ids;
+    fn string(&mut self) -> Result<String, failure::Error>;
+}
+
+impl<R: Read> ReadExt for R {
```

Comment:
I wonder if it'd be a bit cleaner to expose these methods as `Decode for TheType`? For example the `Vec` method could be `Decode for Vec<T>` which albeit not as generic may still suffice? `uleb` could be something like `Decode for u32` or something like that too.

Reply:
This would work for `u32` but not for `Vec<T>` since we don't always want to allocate a vec of the things we deserialize. I'll play with it a bit.